### PR TITLE
Core: Doxygen: Constructors, destructors, operators, and context parameter

### LINKED ISSUE #2298: Docs: Apply common wording
Extracted from #2260:

* Construtors, destructors, operators (including deleted and defaulted)
* Context parameter
* Boolean returns

"Add commonly used template description for all simple/defaulted virtual dtors that give the reason for why we have it."
"Have common phrasing as a starting point across all ctor/dtors for consistency. Constructs the <type name>.../Destructs the <type name> ..."
"Document delete functions like the default ctor, using a template statement."
"For bool returns, follow C, and do ; otherwise, false."

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #2298.

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/base64.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/base64.hpp:25]: What does the `%` symbol do and can you share how this renders in doxygen?

File: sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp:63]: In this context (no pun intended), for getting an access token, is there a request being sent whose lifetime is actually being controlled by the context, or does it server a different purpose?

File: sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp:92]: 👍

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: Just to confirm, isn't this a breaking change, it needs to be called out in the CL, and please try to avoid such subtle changes in large doc updates.

Why do we need to make this change? Please revert changing from public to protected in this PR and open a separate PR for this, and let's discuss the reasons/merits there.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: Why no `%` here? Can you please help me understand the difference.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/datetime.hpp:125]: @vhvb1989 is this extra new line required still?

File: sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp:59]: Same here, @vhvb1989 is this extra new line required still?
I thought we didn't need it anymore. @antkmsft  please confirm and check if we are consistent elsewhere (it was modified recently). I think removing this extra line is necessary to get rid of the "more..." link.

File: sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp:65]: Nice!

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
I was about to block this PR because of the subtle breaking change :(

File: sdk/core/azure-core/inc/azure/core/exception.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/exception.hpp:68]: Consistency nit (consider fixing elsewhere too):
```suggestion
     * @brief Constructs a new `%RequestFailedException` with a \p message string.
```

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:98]: Why did we get rid of this?

File: sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp:59]: @vhvb1989, should we ever have them, or only sometimes, or never?

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:217]: This is a bit confusing to parse.

Maybe modify as something like this:
`...representing the next policy to execute in the sequence of policies.`

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:232]: Good catch!

File: sdk/core/azure-core/inc/azure/core/nullable.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/nullable.hpp:86]: Great comment, containing an important detail.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: Without `%`, it inserts a link. They render just as you'd expect - without the `%` and not as hyperlink. You can look in the build artifacts - https://dev.azure.com/azure-sdk/public/_build/results?buildId=922992&view=artifacts&pathAsName=false&type=publishedArtifacts

File: sdk/core/azure-core/inc/azure/core/response.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/response.hpp:34]: Does this show the templated type correctly in doxygen, like `Response<T>`?

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: Since the instances of the class can't be created because it is abstract, only the classes deriving from it will have access to this constructor, for which `protected` is sufficient.

**Comment by ahsonkhan** [comment]
Other than the code breaking changes which I need to understand the reasoning behind, mostly nits/small suggestions to fix up (and the extra new line).

File: sdk/core/azure-core/inc/azure/core/base64.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/base64.hpp:25]: https://www.doxygen.nl/manual/autolink.html#linkclass

File: sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp:63]: Yes, the request is being sent. Credential is client. Other clients give you files (storage) or secrets (keyvault). Credential classes retreive auth tokens from the server. This is an oversimplification, to an extent.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: It should be there, I missed it!

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:98]: If custom constructor is present, default constructor does not get generated either way.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:98]: But I think later we should explicitly list and document all the default operations for all the code. We have a work item for that. We currently don't do that, and it was not worth it to document the deletion of the constructor that happens by the C++ rules anyway.

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:217]: I think to combine would be better "`is an abstraction representing the next policy in the stack sequence of policies, from the caller's perspective.`".

File: sdk/core/azure-core/inc/azure/core/response.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: It would print the name in monospace font without the link, i.e without the self-reference.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: Will do a follow up PR.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:125]: Maybe will be in a follow up PR, depending on the answer

File: sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp:59]: Maybe will be in a follow up PR, depending on the answer

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: I think we should leave it, it is not technically a breaking change, since you can't call a public constructor of an abstract class, other than deriving from it, for which `protected` constructor is accessible.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:59]: Will take to the follow up PR.

File: sdk/core/azure-core/inc/azure/core/exception.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/exception.hpp:68]: Will take to the follow up PR.

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:217]: Will bring in the follow up PR.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:98]: I am planning to not take this to the follow up PR.

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: I am planning to not take this to the follow up PR.

File: sdk/core/azure-core/inc/azure/core/response.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: Please LMK if you think we should be monospacing it as `Response<T>` (then it should apply to `Nullable<T>` as well). I don't think we should.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: So far, not taking this to a follow up PR.

File: sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp:59]: We want it for docs inside a class, like properties and methods.

We don't want it for documentation out of classes, like struct/class/namespace/global stuff

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/diagnostics/logger.hpp:59]: Victor: "We want it for docs inside a class, like properties and methods.

We don't want it for documentation out of classes, like struct/class/namespace/global stuff"

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:125]: Victor "We want it for docs inside a class, like properties and methods.

We don't want it for documentation out of classes, like struct/class/namespace/global stuff"

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:217]: What do you mean by "stack sequence of policies" and what does "the caller's perspective" mean? Those two phrases confuse me a bit, so I question their usefulness. Your call of course :)

File: sdk/core/azure-core/inc/azure/core/response.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: > monospacing it as `Response<T>`

What do you mean by that?

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: #2390

File: sdk/core/azure-core/inc/azure/core/exception.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/exception.hpp:68]: #2390

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:98]: @CaseyCarter for the extensible enum pattern, should we explicitly delete the default ctor or leave it to the compiler to generate?
The guideline doesn't have it:
https://azure.github.io/azure-sdk/cpp_implementation.html#enumerations

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:217]: Maybe I am biased, because I think I know exactly what this means, and it makes perfect sense to me.
"next" policy is in relation to who? To the caller. I.e. whenever you see this and yo can call it, it is next relative to you, not in the absolute sense.
And as of the stack. Policies are a stack.

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: Do we need to define a ctor at all in this case? For example, BodyStream is an abstract class, that doesn't have a ctor.

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:217]: I like the wording that I put in ##2378. If you give it to me, I'll leave it as I have it in that PR. But I don't feel super strongly on this, if you think of a better wording, please propose it in that PR.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:59]: #2390

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: You are right - we don't need to (we would've need if if we needed a custom one, but since we are using the default - we don't need to define it at all

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: I think a few other instances were missed. For example:
https://github.com/antkmsft/azure-sdk-for-cpp/blob/6185cba9e6701a1943f48f52963c963637cd8e52/sdk/core/azure-core/inc/azure/core/datetime.hpp#L52

https://github.com/antkmsft/azure-sdk-for-cpp/blob/6185cba9e6701a1943f48f52963c963637cd8e52/sdk/core/azure-core/inc/azure/core/datetime.hpp#L134-L136

https://github.com/antkmsft/azure-sdk-for-cpp/blob/6185cba9e6701a1943f48f52963c963637cd8e52/sdk/core/azure-core/inc/azure/core/datetime.hpp#L172

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: Or, if there were custom constructors with parameters, but we also wanted to have a parameterless one.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:50]: I added that to #2390.

File: sdk/core/azure-core/inc/azure/core/response.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: Nevermind, I get what you mean by `monospacing`, the "code" like font. I was more curious about the template `<T>`, maybe we should explicitly add that whenever referring to the type in the docs. So, not `Response` but rather `Response<T>`

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: Not fixing all of them so far, only the places I was adding. There are methods next to constructors, and they do still do self-reference. There are work items to go and update classes in each namespace, divided by the team member, it can be covered there. I'll go and update comments there.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:117]: I updated #2299, #2300, and #2304

File: sdk/core/azure-core/inc/azure/core/response.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: See how it renders in the published artifacts - https://dev.azure.com/azure-sdk/public/_build/results?buildId=922992&view=artifacts&pathAsName=false&type=publishedArtifacts

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: I.e. 
image

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: > > I don't think we should.
> I am curious, how come?

Doxygen refers to it as simply `Response`, not `Response<>` or `Response<T>`. `T` is also a type parameter name, although a very common one. What if we rename `T` to `ResponseType`, should we start referring to it as `Response<ResponseType>`? I think of `<T>` as an implementation detail. It could've been used to distinguish between different `Response` types, should've had a templated and not templated versions, but this is not the case.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: And if we'd have to emphasize it is a template, I think more pedantic way to call it would be `Response<>`, not `Response<T>`.

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/response.hpp:34]: Makes sense, I think we can leave it as you have it for now.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp
Comment (CaseyCarter) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:98]: I'm perfectly happy for the default constructor to be implicitly deleted without calling it out explicitly. This isn't as likely to cause confusion as are the arcane rules around when move special members are generated or suppressed which motivates the rule of 0/3/5 guideline.

**Pull request merged**
